### PR TITLE
[RLlib] `Trainer` to `Algorithm` renaming

### PR DESCRIPTION
**Pull request by sven1977**
* Rename class `Trainer` into `Algorithm`.
* Rename class `TrainerConfig` into `AlgorithmConfig`. Same for `(Partial)?TrainerConfigDict` (soon to be deprecated as we are moving to AlgorithmConfig objects, away from python config dicts).
* Replace (almost all) mentions of `[Tt]rainer` in code and docs with `[Aa]lgorithm`/`algo`.
* Add backward compat tests, making sure that ray 1.10 imports from the old `agents` directory still work and that all old 1.10 config dict keys are still supported (only create deprecation warnings).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kouroshHakha** [comment]
Do we want to move the core files (i.e `algorithm.py, algorithm_config.py, callbacks.py, registry.py`, and `mock.py`) to `rllib.core` or keep them under the algorithms folder?

File: rllib/agents/__init__.py [hunk 1]
```diff
@@ -1,16 +1,8 @@
-from ray.rllib.agents.callbacks import (
-    DefaultCallbacks,
-    MemoryTrackingCallbacks,
-    MultiCallbacks,
-)
-from ray.rllib.agents.trainer import Trainer, with_common_config
-from ray.rllib.agents.trainer_config import TrainerConfig
```
Comment (kouroshHakha) [inline rllib/agents/__init__.py:7]: This statement is not backward compatible. We need deprecation of all core files under `algorithms.py`

File: rllib/__init__.py [hunk 2]
```diff
@@ -28,8 +28,8 @@ def _setup_logger():
 
 
 def _register_all():
-    from ray.rllib.agents.trainer import Trainer
-    from ray.rllib.agents.registry import ALGORITHMS, get_trainer_class
```
Comment (kouroshHakha) [inline rllib/__init__.py:32]: Lost short-term backward compatibility.

**Comment by kouroshHakha** [comment]
One major problem right now is that we don't have nice deprecation of anything other than the `Trainer` class. Basically trainer_config, mock, registry, .etc they are completely removed in this PR.

**Comment by kouroshHakha** [comment]
@sven1977 This is a monstrous PR, and I am afraid that it may break backward compatibility without us noticing it. Do you think we can break it down to multiple stages? My suggestion is the following: 

1. renaming core pieces first (trainer, trainer_config, etc.) and making sure the old algorithm tests still pass by referencing the old classes while getting right deprecation messages?
2. update all algorithms without their tests, the tests should still pass without breaking (may only throw some deprecation warnings)
3. update algorithm tests --> This should remove all deprecation warnings within algorithm test
4. update the docs
5. update the examples and the general rllib tests --> This should remove all deprecation warnings within all rllib tests
6. update everything else that remained
7. update in-file doc-strings
8. Things like should we update public method names like `merge_trainer_config`?

Basically at non of these stages should we skip checking testing backward compatibility of any kind.

File: rllib/tests/backward_compat/old_ppo.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+from ray.rllib.agents import with_common_config
```
Comment (kouroshHakha) [inline rllib/tests/backward_compat/old_ppo.py:1]: Why name this old_ppo.py?

File: rllib/tests/backward_compat/test_backward_compat.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            + list(CONTRIBUTED_ALGORITHMS.keys())
+            + ["__fake", "__sigmoid_fake_data", "__parameter_tuning"]
+        ):
+            _get_trainer_class(key)
+
+        # Also test this for all `CONTRIBUTED_ALGORITHMS`.
+        for key in list(CONTRIBUTED_ALGORITHMS.keys()):
+            _get_trainer_class(key)
+
+    def test_old_configs(self):
```
Comment (kouroshHakha) [inline rllib/tests/backward_compat/test_backward_compat.py:35]: Don't you wanna do this for all algos in the combined list of `ALGORITHMS` and `CONTRIBUTED_ALGORITHMS` from agents / contrib? why only ppo?

File: rllib/tests/backward_compat/old_ppo.py (hunk 3 above)
Comment (sven1977) [reply rllib/tests/backward_compat/old_ppo.py:1]: Well, it's the old PPO python config dict.

File: rllib/tests/backward_compat/test_backward_compat.py (hunk 4 above)
Comment (sven1977) [reply rllib/tests/backward_compat/test_backward_compat.py:35]: I think this would blow up the test significantly if we tested ALL algos in here. The main point is to test the generic (shared) Trainer settings with one example algo.

File: rllib/__init__.py (hunk 2 above)
Comment (sven1977) [reply rllib/__init__.py:32]: There is a new test for this now that uses all the "old way" import statements based on older registry code.

File: rllib/agents/__init__.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/__init__.py:7]: See below:
At least importing all Algorithms (under the old "...Trainer" name) AND their default configs (under their old python dict name) is tested now via a new test case that uses the old registry code.

**Review by kouroshHakha — APPROVED** [review]

**Review by maxpumperla — APPROVED** [review]
